iroh-gateway/src/core.rs

```diff
@@ -298,87 +426,167 @@ mod tests {
+            None,
+        )
+        .await;
 
-        // request the root cid as a recursive car
-        let res = {
-            let client = hyper::Client::new();
-            let uri = hyper::Uri::builder()
-                .scheme("http")
-                .authority(format!("localhost:{}", addr.port()))
-                .path_and_query(format!("/ipfs/{}/{}", root_cid, "world.txt"))
-                .build()
-                .unwrap();
-            let req = hyper::Request::builder()
-                .method("HEAD")
-                .uri(uri)
-                .body(hyper::Body::empty())
-                .unwrap();
-            client.request(req).await.unwrap()
-        };
+        assert_eq!(http::StatusCode::MOVED_PERMANENTLY, res.status());
+        assert_eq!(
+            format!(
+                "http://{}.ipfs.localhost:{}/",
+                test_setup.root_cid,
+                test_setup.gateway_addr.port()
+            ),
+            res.headers().get("Location").unwrap().to_str().unwrap()
+        );
 
-        assert_eq!(http::StatusCode::OK, res.status());
-        assert!(res.headers().get("content-length").is_some());
-        assert_eq!(res.headers().get("content-length").unwrap(), "5");
+        // IPNS
+        let res = do_request(
+            "GET",
+            &format!("localhost:{}", test_setup.gateway_addr.port()),
+            "/ipns/k51qzi5uqu5dlvj2baxnqndepeb86cbk3ng7n3i46uzyxzyqj2xjonzllnv0v8",
+            None,
+        )
+        .await;
 
-        let (body, _) = res.into_body().into_future().await;
-        assert!(body.is_none());
+        assert_eq!(http::StatusCode::MOVED_PERMANENTLY, res.status());
+        assert_eq!(
+            format!(
+                "http://k51qzi5uqu5dlvj2baxnqndepeb86cbk3ng7n3i46uzyxzyqj2xjonzllnv0v8.ipns.localhost:{}/",
+                test_setup.gateway_addr.port()
+            ),
+            res.headers().get("Location").unwrap().to_str().unwrap()
+        );
 
-        core_task.abort();
-        core_task.await.unwrap_err();
-        store_task.abort();
-        store_task.await.unwrap_err();
+        // Test that IPNS records are recoded to base36
+        let res = do_request(
+            "GET",
+            &format!("localhost:{}", test_setup.gateway_addr.port()),
+            "/ipns/bafyreihyrpefhacm6kkp4ql6j6udakdit7g3dmkzfriqfykhjw6cad5lrm",
+            None,
+        )
+        .await;
+
+        assert_eq!(http::StatusCode::MOVED_PERMANENTLY, res.status());
+        assert_eq!(
+            format!(
+                "http://k2jvslbl2n1p4suo7yr973y3v7pfautpxba2jeb9fpmjil0l3lppcgi3.ipns.localhost:{}/",
+                test_setup.gateway_addr.port()
+            ),
+            res.headers().get("Location").unwrap().to_str().unwrap()
+        );
+
+        // IPNS + DNSLink
+        let res = do_request(
+            "GET",
+            &format!("localhost:{}", test_setup.gateway_addr.port()),
+            "/ipns/ipfs-subdomain.ipfs-domain.eth",
```

Comment:
nit: change to `en.wikipedia-on-ipfs.org`? this is our usual test case for handling `-`  (i suggest avoiding using  .eth in tests unrelated to ENS interop).

Reply:
Done.

Btw, while I was implementing code around `-` processing with regular expressions I had faced that it is hard or even impossible to properly process cases like `aaaa-b-aaaa.com` (a single symbol between 2 dashes) because captured groups were overlapping there (`a-b` and `b-a` with `b` belonging to both groups) and it is forbidden in `Regex` crate.

I mean, it may be useful to consider in E2E and other tests this case too cause it belongs to a separate equivalence class.

Reply:
fwiw Kubo also started with regex, realized it comes with footguns, and replaced with a plain function :)

Reply:
And it doesn't bring extra crate. You are right, dunno why I've decided regex would be better. It might be useful for complicated cases or expected future changes in format, but replacing dashes are definitely not this case.